File: ui_tests/helpers/smart_status_helper.js
// Helper to poll /batch/api/smart_processing_status from Playwright JS tests.
// Usage:
// const { pollSmartProcessingStatus } = require('./helpers/smart_status_helper');
// const { lastEvent, meta } = await pollSmartProcessingStatus(page, token, { baseUrl });

async function pollSmartProcessingStatus(pageOrRequest, token, opts = {}) {
  const baseUrl = opts.baseUrl || process.env.BASE_URL || 'http://127.0.0.1:5000'
  const maxPolls = opts.maxPolls || 60
  const stallLimit = opts.stallLimit || 10
  const pollInterval = opts.pollInterval || 1000

  // pageOrRequest may be a Playwright `page` or `APIRequestContext` (page.request)
  const requestCtx = pageOrRequest.request ? pageOrRequest.request : pageOrRequest

  let lastProgress = null
  let stallCount = 0
  let lastEvent = null

  for (let i = 0; i < maxPolls; i++) {
    try {
      const resp = await requestCtx.get(`${baseUrl}/batch/api/smart_processing_status?token=${encodeURIComponent(token)}`)
      const body = await resp.json()
      lastEvent = body?.data?.last_event ?? null
      if (lastEvent && typeof lastEvent === 'object') {
        const prog = lastEvent.progress
        const complete = Boolean(lastEvent.complete)
        if (typeof prog !== 'undefined' && prog !== null) {
          if (prog !== lastProgress) {
            lastProgress = prog
            stallCount = 0
          } else {
            stallCount += 1
          }
        } else {
          stallCount += 1
        }
        if (complete) {
          return { lastEvent, meta: { polls: i + 1, stalled: false, completed: true, lastProgress } }
        }
        if (stallCount >= stallLimit) {
          return { lastEvent, meta: { polls: i + 1, stalled: true, completed: false, lastProgress } }
        }
      }
    } catch (e) {
      // ignore and retry
    }
    await new Promise(r => setTimeout(r, pollInterval))
  }
  return { lastEvent, meta: { polls: maxPolls, stalled: stallCount >= stallLimit, completed: false, lastProgress } }
}

module.exports = { pollSmartProcessingStatus }

File: doc_processor/tests/e2e/smart_status_helper.py
"""Helper utilities for polling smart processing fallback status in tests.

Provides `poll_smart_processing_status` which repeatedly polls the
`/batch/api/smart_processing_status` endpoint until processing completes
or progress stalls.
"""
import time
import requests
from typing import Any, Dict, Optional, Tuple


def poll_smart_processing_status(token: str, base_url: str = 'http://127.0.0.1:5000',
                                 max_polls: int = 60, stall_limit: int = 10,
                                 poll_interval: float = 1.0, timeout: float = 5.0,
                                 session: Optional[requests.Session] = None) -> Tuple[Optional[Dict[str, Any]], Dict[str, Any]]:
    """Poll the smart processing fallback endpoint until complete or stalled.

    Returns a tuple `(last_event, meta)` where `last_event` is the last
    `last_event` dict returned by the server (or None) and `meta` contains
    diagnostic info: `polls`, `stalled`, `completed`, `last_progress`.
    """
    status_url = f"{base_url.rstrip('/')}/batch/api/smart_processing_status?token={token}"
    s = session or requests.Session()
    last_progress = None
    stall_count = 0
    polls = 0
    for attempt in range(max_polls):
        polls += 1
        try:
            r = s.get(status_url, timeout=timeout)
        except Exception as e:
            # Return diagnostic info but continue polling until max
            time.sleep(poll_interval)
            continue
        try:
            j = r.json()
        except Exception:
            j = {}
        last = j.get('data', {}).get('last_event') if isinstance(j, dict) else None
        if isinstance(last, dict):
            prog = last.get('progress')
            complete = bool(last.get('complete'))
            if prog is not None:
                if prog != last_progress:
                    last_progress = prog
                    stall_count = 0
                else:
                    stall_count += 1
            else:
                stall_count += 1

            if complete:
                return last, {'polls': polls, 'stalled': False, 'completed': True, 'last_progress': last_progress}

            if stall_count >= stall_limit:
                return last, {'polls': polls, 'stalled': True, 'completed': False, 'last_progress': last_progress}
        time.sleep(poll_interval)
    return last if 'last' in locals() else None, {'polls': polls, 'stalled': stall_count >= stall_limit, 'completed': False, 'last_progress': last_progress}
